text/0000-float-semantics.md

```diff
@@ -0,0 +1,305 @@
+This means that even though arithmetic operations can produce signaling NaNs, there is a guarantee that signaling NaNs will never appear "out of thin air".
+LLVM does not actually document that they are using IEEE float semantics, but de-facto they do on almost all targets (the exception are targets that use x87 instructions, as noted above).
+
+Java requires exact IEEE 754-2008 compliance and goes through a lot of effort to realize that on 32bit x86 without SSE (see [here](https://open-std.org/jtc1/sc22/jsg/docs/m3/docs/jsgn325.pdf) and [here](https://open-std.org/JTC1/SC22/JSG/docs/m3/docs/jsgn326.pdf)). However, they do not seem to tackle the issue of specifying NaN payload bits, even though those bits [can be observed](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Float.html#floatToRawIntBits(float)).
+
+wasm [guarantees](https://webassembly.github.io/spec/core/exec/numerics.html#nan-propagation) that "if all input NaNs are canonical, then any output NaN is canonical". Here ["canonical"](https://webassembly.github.io/spec/core/syntax/values.html#canonical-nan) is defined as a NaN with a "payload whose most significant bit 1 is while all others are 0", i.e. it matches what we call "preferred" above.
+(We are departing from wasm terminology since "canonical" already has a different meaning in the IEEE spec. For similar reasons, we consider the NaN payload to *not* include the quiet/signaling bit, whereas wasm considers the quiet/signaling bit to be part of the payload.)
+The sign bit is left unspecified, i.e., there are two canonical NaNs (and both are quiet under the standard interpretation of the signaling/quiet bit).
+
+In Rust itself, questions around float semantics have been discussed for a long time.
+[This issue](https://github.com/rust-lang/unsafe-code-guidelines/issues/237) collects a lot of that discussion, which culminated in this RFC.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Should we be concerned that LLVM does [not actually document](https://github.com/llvm/llvm-project/issues/60942) that it uses IEEE float semantics?
+  It does assume IEEE semantics in its own optimization passes.
+- Are there any other targets with floating-point trouble?
+- What exactly is the set of "extra" NaNs for all remaining targets?
+- To what extend does this specification apply to platform intrinsics?
+  On the one hand, it seems reasonable to expect platform intrinsics to have the behavior of the platform instructions.
+  On the other hand, we implement some platform intrinsics with the portable LLVM `simd` intrinsics, and those are subject to the NaN-non-determinism described above.
+  So the current de-facto semantics of at least some platform intrinsics is that they do *not* match what the platform does.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
```

Comment:
We could have a lint that triggers when a const op generates a float.

Reply:
@RalfJung I think you meant "generates a NaN" here?

Reply:
Ah, yes.^^ Fixed.